Address issues from new compiler flags

bgeltz opened the pull request:
- Relates to #3710.

cmcantalupo commented on libgeopmd/src/StatsCollector.cpp:
```diff
@@ -152,13 +152,13 @@ namespace geopm
-                m_stats->mean(metric_idx),
-                m_stats->std(metric_idx),
+                static_cast<double>(m_stats->count(static_cast<int>(metric_idx))),
+                static_cast<double>(m_stats->first(static_cast<int>(metric_idx))),
+                static_cast<double>(m_stats->last(static_cast<int>(metric_idx))),
+                static_cast<double>(m_stats->min(static_cast<int>(metric_idx))),
+                static_cast<double>(m_stats->max(static_cast<int>(metric_idx))),
+                static_cast<double>(m_stats->mean(static_cast<int>(metric_idx))),
+                static_cast<double>(m_stats->std(static_cast<int>(metric_idx))),
```
```suggestion
                static_cast<double>(m_stats->first(static_cast<int>(metric_idx))),
                m_stats->last(static_cast<int>(metric_idx)),
                m_stats->min(static_cast<int>(metric_idx)),
                m_stats->max(static_cast<int>(metric_idx)),
                m_stats->mean(static_cast<int>(metric_idx)),
                m_stats->std(static_cast<int>(metric_idx)),
```

cmcantalupo commented on libgeopmd/src/LevelZeroDevicePool.cpp:
```diff
@@ -180,9 +180,9 @@ namespace geopm
-                                          (supported_frequency.size() - 1);
+            frequency_step_mhz = static_cast<double>(supported_frequency.back() -
+                                                     supported_frequency.front()) /
+                                  static_cast<double>(supported_frequency.size() - 1);
```
ws?

cmcantalupo commented on libgeopmd/src/SysfsIOGroup.cpp:
```diff
@@ -52,7 +52,7 @@ namespace geopm
     {
         char buf[SysfsDriver::M_IO_BUFFER_SIZE] = {0};
-        int read_bytes = read(fd, buf, sizeof buf - 1);
+        int read_bytes = static_cast<int>(read(fd, buf, sizeof buf - 1));
```
```suggestion
        auto read_bytes = read(fd, buf, sizeof buf - 1);
```
then you can forgo teh static cast on line 60.  Not really a big deal

bgeltz replied:
Read returns a ssize_t type (signed).  sizeof (in C++11) returns a std::size_t type (unsigned).  I'm going to have to cast these both to keep the logic we have.

cmcantalupo commented on libgeopmd/src/SysfsIOGroup.cpp:
```diff
@@ -69,7 +69,7 @@ namespace geopm
     static void write_resource_attribute_fd(int fd, const std::string &value)
     {
-        int write_bytes = pwrite(fd, value.c_str(), value.length() + 1, 0);
+        int write_bytes = static_cast<int>(pwrite(fd, value.c_str(), value.length() + 1, 0));
```
same pattern as above...

bgeltz replied:
It's the same problem here:

pwrite returns a ssize_t type (signed).  Line 77's RHS is a std::size_t from value.length().

cmcantalupo reviewed (commented):
Couple of comments.

bgeltz committed:
Add new compiler flags

- Add Wimplicit-fallthrough compiler flag.
- Add CPP flag for _GLIBCXX_ASSERTIONS.
- Add stack-protector and clash protection flags.

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

cmcantalupo commented on libgeopm/configure.ac:
```diff
@@ -550,9 +552,9 @@ AC_COMPILE_IFELSE([
-[LDFLAGS_NOPIE="$AM_LDFLAGS -Wl,-z,relro,-z,now,-z,noexecstack"]
+[AM_CFLAGS="$AM_CFLAGS -Wall -Wextra -Werror -Wno-unused-parameter -Wformat -Wformat-security -Wimplicit-fallthrough $NO_FASTMATH -fPIC -fPIE -fstack-protector-strong -fstack-clash-protection -include config.h"]
+[AM_CXXFLAGS="$AM_CXXFLAGS -Wall -Wextra -Werror -Wno-unused-parameter -Wformat -Wformat-security -Wimplicit-fallthrough $NO_FASTMATH -fPIC -fPIE -fstack-protector-strong -fstack-clash-protection -include config.h"]
+[LDFLAGS_NOPIE="$AM_LDFLAGS -Wl,-z,relro,-z,now,-z,noexecstack,-z,nodlopen"]
```
This seems to be breaking cffi, which of course needs to dlopen libgeopmd and libgeopm

bgeltz committed:
Typos and config clarification

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

bgeltz committed:
Add rust flags

Signed-off-by: Brad Geltz <brad.geltz@intel.com>
